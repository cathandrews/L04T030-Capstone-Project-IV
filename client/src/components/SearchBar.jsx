import { useState } from "react";
import styles from "./styles/SearchBar.module.css";

/**
 * SearchBar Component
 *
 * A reusable search input component that allows users to enter a search query.
 * When the form is submitted, it calls the `onSearch` function with the current query.
 *
 * @param {Function} onSearch - Callback function to handle the search query.
 */
const SearchBar = ({ onSearch }) => {
  // Local state to store the current input value
  const [query, setQuery] = useState("");

  /**
   * Handle form submission
   * Prevents default form behavior and calls the onSearch callback
   * with the current query value
   */
  const handleSubmit = (e) => {
    e.preventDefault();
    onSearch(query);
  };

  return (
    <form onSubmit={handleSubmit} className={styles.form}>
      {/* Input field bound to the query state */}
      <input
        type="text"
        value={query}
        onChange={(e) => setQuery(e.target.value)}
        placeholder="Search users..."
        className={styles.input}
      />
      {/* Submit button triggers handleSubmit */}
      <button type="submit" className={styles.button}>
        Search
      </button>
    </form>
  );
};

export default SearchBar;
