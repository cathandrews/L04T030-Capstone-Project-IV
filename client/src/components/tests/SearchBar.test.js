import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import SearchBar from "../SearchBar";

describe("SearchBar Component", () => {
  // Mock function to simulate the onSearch callback
  const mockOnSearch = jest.fn();

  beforeEach(() => {
    // Clear all mock function calls before each test
    mockOnSearch.mockClear();
  });

  // Test 1: Check if the component renders correctly
  it("renders correctly", () => {
    render(<SearchBar onSearch={mockOnSearch} />);
    expect(screen.getByPlaceholderText("Search users...")).toBeInTheDocument();
    expect(screen.getByText("Search")).toBeInTheDocument();
  });

  // Test 2: Check if the query state updates when input changes
  it("updates the query state when input changes", () => {
    render(<SearchBar onSearch={mockOnSearch} />);
    const input = screen.getByPlaceholderText("Search users...");
    fireEvent.change(input, { target: { value: "test" } });
    expect(input.value).toBe("test");
  });

  // Test 3: Check if onSearch is called with the correct value when form is submitted
  it("calls onSearch with the input value when form is submitted", () => {
    render(<SearchBar onSearch={mockOnSearch} />);
    const input = screen.getByPlaceholderText("Search users...");
    const button = screen.getByText("Search");
    fireEvent.change(input, { target: { value: "test" } });
    fireEvent.click(button);
    expect(mockOnSearch).toHaveBeenCalledWith("test");
  });

  // Test 4: Snapshot test using @testing-library/react
  it("matches snapshot", () => {
    // Render the component
    const { asFragment } = render(<SearchBar onSearch={mockOnSearch} />);
    // Use the rendered output for the snapshot
    expect(asFragment()).toMatchSnapshot();
  });
});
