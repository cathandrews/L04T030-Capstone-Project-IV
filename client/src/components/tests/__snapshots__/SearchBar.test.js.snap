// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar Component matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="form"
  >
    <input
      class="input"
      placeholder="Search users..."
      type="text"
      value=""
    />
    <button
      class="button"
      type="submit"
    >
      Search
    </button>
  </form>
</DocumentFragment>
`;
